synapseclient/models/services/storable_entity_components.py

```diff
@@ -0,0 +1,143 @@
+import asyncio
+from typing import Union, TYPE_CHECKING, Optional
+from synapseclient import Synapse
+from synapseclient.models import Annotations
+
+if TYPE_CHECKING:
+    from synapseclient.models import File, Folder, Project, Table
+
+
+async def store_entity_components(
```

Comment:
Currently the behavior is that when an exception is raised (See the sample script below) the futures that are yet to be executed as cancelled and do not finish. This means it is as you are saying: "If upload fails for file 003 of the 10,000 files we raise the exception and cancel the rest of the upload".

```
import asyncio
import os
import uuid

from synapseclient.models import File, Folder
from datetime import date, datetime, timedelta, timezone
import synapseclient

PROJECT_ID = "syn52948289"

syn = synapseclient.Synapse(debug=True)
syn.login()


def create_random_file(
    path: str,
) -> None:
    """Create a random file with random data.

    :param path: The path to create the file at.
    """
    with open(path, "wb") as f:
        f.write(os.urandom(1))


async def store_file():
    # Cleanup synapse for previous runs - Does not delete local files/directories:
    script_file_folder = Folder(name="file_script_folder", parent_id=PROJECT_ID)
    if not os.path.exists(os.path.expanduser("~/temp/myNewFolder")):
        os.mkdir(os.path.expanduser("~/temp/myNewFolder"))
    # Hack to get the ID as Folder does not support get by name/id yet
    await script_file_folder.store()
    await script_file_folder.delete()
    await script_file_folder.store()

    # Creating annotations for my file ==================================================
    annotations_for_my_file = {
        "my_single_key_string": "a",
        "my_key_string": ["b", "a", "c"],
        "my_key_bool": [False, False, False],
        "my_key_double": [1.2, 3.4, 5.6],
        "my_key_long": [1, 2, 3],
        "my_key_date": [date.today(), date.today() - timedelta(days=1)],
        "my_key_datetime": [
            datetime.today(),
            datetime.today() - timedelta(days=1),
            datetime.now(tz=timezone(timedelta(hours=-5))),
            datetime(2023, 12, 7, 13, 0, 0, tzinfo=timezone(timedelta(hours=0))),
            datetime(2023, 12, 7, 13, 0, 0, tzinfo=timezone(timedelta(hours=-7))),
        ],
    }

    # 1. Creating a file =================================================================
    files_to_upload = []
    for i in range(10):
        name_of_file = f"file_script_my_file_with_random_data_{uuid.uuid4()}.txt"
        path_to_file = os.path.join(os.path.expanduser("~/temp"), name_of_file)
        create_random_file(path_to_file)
        file = File(
            path=path_to_file,
            annotations=annotations_for_my_file,
            parent_id=script_file_folder.id,
            description="This is a file with random data.",
        )
        files_to_upload.append(file)
    files_to_upload.append(
        File(
            path="~/temp/myNewFolder/file_that_does_not_exist.txt",
            parent_id=script_file_folder.id,
        )
    )

    script_sub_folder = Folder(
        name="file_script_sub_folder", parent_id=script_file_folder.id
    )
    script_sub_folder.files = files_to_upload
    await script_sub_folder.store()


asyncio.run(store_file())
```

Are you thinking that instead of this the default behavior is that we log the exception/stack trace, but let the rest of the program execute till it finishes? Do you have any thoughts about the behavior in this case as it relates to filling back in the failed `File` instance? If 10k files are being uploaded and 100 fail I don't want to sift through the console logs as the only thing to tell me that "These are the files that failed". My initial (Not thought out) solution is to stash an attribute on the `File` that notes the last interaction with Synapse. Something like:

```
InteractionStatus:
    successful: bool
    exception: Exception
```

Reply:
cc @BWMac @jaymedina @danlu1 Any thoughts on the behavior of this piece?

Reply:
I agree that it would be annoying if one failure of 10000 prevents the rest of the uploads. This is a pattern that we often encounter with Nextflow Workflows. 

My initial thoughts are to do something like @BryanFauble suggests with a status updated upon success/failure, or to have some way to optionally return the files that failed to be uploaded in a data structure. The latter would enable a user to re-try in the same script by iterating back through the failed files and you could decide how many times you want to re-try before you give up on those persistently failing uploads.

Reply:
@BWMac @thomasyu888 
Maybe for methods that kick off bulk actions (Think `Project.store()`, `Folder.store()`) we have an argument you can pass into store like:

```
class FailureStrategy(Enum):
    """
    When storing a large number of items through bulk actions like `Project.store()` or 
    `Folder.store()` individual failures may occur. Passing this ENUM will allow you to 
    define how you want to respond to failures.
    """

    RAISE_EXCEPTION = "RAISE_EXCEPTION"
    """An exception is raised on the first failure and all tasks yet to be completed 
    are cancelled."""
    
    LOG_EXCEPTION = "LOG_EXCEPTION"
    """An exception is logged and all tasks yet to be completed continue to be 
    processed."""
    
    RETURN_INTERACTION_FAILURE = "RETURN_INTERACTION_FAILURE"
    """
    When a File or Folder fails to be stored in Synapse a `synchronization_failure` 
    exception will be stored on the failed File or Folder instance. Use the `.failures` 
    property to retrieve a generator that yields a named tuple of:
    {
        "exception": Exception,
        "entity": Union[File, Folder],
        "parent": Union[Project, Folder],
    }
    """

    CALLBACK = "CALLBACK"
    """
    Call back to a user defined function when a failure occurs. The callback function
    should take in a single argument which will be a named tuple of:
    {
        "exception": Exception,
        "entity": Union[File, Folder],
        "parent": Union[Project, Folder],
    }
    """

```

Reply:
Also - In a first iteration we could also provide only `RAISE_EXCEPTION` or `LOG_EXCEPTION`

Reply:
I like this idea, we can add a ticket for this.

Reply:
I'll plan to include this with the changes in https://sagebionetworks.jira.com/browse/SYNPY-1415